# Korrelation und Regression

## World Happiness Report

Der World Happiness Report (https://worldhappiness.report/) ist ein jährlich vom Sustainable Development Solutions Network der Vereinten Nationen veröffentlichter Bericht. Der Bericht enthält Ranglisten zur Lebenszufriedenheit in verschiedenen Ländern der Welt und Datenanalysen aus verschiedenen Perspektiven. 

Die wichtigste Frage der Umfrage, auf der der Bericht basiert, ist die so genannte *Cantril-Leiter*: Die Befragten werden gebeten, sich eine Leiter vorzustellen, bei der das bestmögliche Leben für sie eine 10 und das schlechtestmögliche Leben eine 0 ist. Anschließend werden sie gebeten, ihr eigenes aktuelles Leben auf dieser Skala von 0 bis 10 zu bewerten.

## Die Daten

Wir laden den Bericht aus dem Jahr 2022.

In [1]:
import pandas as pd
import plotly.express as px

whr = pd.read_csv("Daten/WHR2022.csv")

### ✏️ Aufgabe 1
Welche Dimension haben die Daten? Was repräsentiert eine Zeile in den Daten? 
Was vermuten Sie repräsentieren die Spalten? Details können Sie in der Datei WHR+22.pdf auf S.15 und S.21 nachschlagen!

In [2]:
whr.shape

(2089, 12)

In [3]:
whr

Country name  year  Life Ladder  Log GDP per capita  Social support  \
0     Afghanistan  2008     3.723590            7.302574        0.450662   
1     Afghanistan  2009     4.401778            7.472446        0.552308   
2     Afghanistan  2010     4.758381            7.579183        0.539075   
3     Afghanistan  2011     3.831719            7.552006        0.521104   
4     Afghanistan  2012     3.782938            7.637953        0.520637   
...           ...   ...          ...                 ...             ...   
2084     Zimbabwe  2017     3.638300            8.241609        0.754147   
2085     Zimbabwe  2018     3.616480            8.274620        0.775388   
2086     Zimbabwe  2019     2.693523            8.196998        0.759162   
2087     Zimbabwe  2020     3.159802            8.117733        0.717243   
2088     Zimbabwe  2021     3.154578            8.153248        0.685151   

      Healthy life expectancy at birth  Freedom to make life choices  \
0                            50.500000                      0.718114   
1                            50.799999                      0.678896   
2                            51.099998                      0.600127   
3                            51.400002                      0.495901   
4                            51.700001                      0.530935   
...                                ...                           ...   
2084                         52.150002                      0.752826   
2085                         52.625000                      0.762675   
2086                         53.099998                      0.631908   
2087                         53.575001                      0.643303   
2088                         54.049999                      0.667636   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  \
0       0.173169                   0.881686         0.414297         0.258195   
1       0.195469                   0.850035         0.481421         0.237092   
2       0.125859                   0.706766         0.516907         0.275324   
3       0.167723                   0.731109         0.479835         0.267175   
4       0.241247                   0.775620         0.613513         0.267919   
...          ...                        ...              ...              ...   
2084   -0.113937                   0.751208         0.733641         0.224051   
2085   -0.084747                   0.844209         0.657524         0.211726   
2086   -0.081540                   0.830652         0.658434         0.235354   
2087   -0.029376                   0.788523         0.660658         0.345736   
2088   -0.109439                   0.756945         0.609917         0.241682   

      Confidence in national government  
0                              0.612072  
1                              0.611545  
2                              0.299357  
3                              0.307386  
4                              0.435440  
...                                 ...  
2084                           0.682647  
2085                           0.550508  
2086                           0.456455  
2087                           0.577302  
2088                           0.665703  

[2089 rows x 12 columns]

Eine Zeile stellt ein ganzes Land dar

### ✏️ Aufgabe 2
Der Datensatz erstreckt sich über viele Jahre, aber wir sind nur an den Daten aus dem Jahr 2021 interessiert. Filtern Sie den DataFrame `whr` für das Jahr 2021, wählen Sie alle Spalten aus und weisen Sie das Ergebnis einem DataFrame mit dem Namen `whr_2021` zu.

In [4]:
whr_2021 = whr[whr.year == 2021]

### ✏️ Aufgabe 3
Wie hoch ist die durchschnittliche Lebenserwartung `Healthy life expectancy at birth` über alle Länder? Können Sie aus den Daten auch die durchschnittliche Lebenserwartung aller Menschen berechnen? (wenn ja: wie? wenn nein: warum nicht?)

In [5]:
whr.columns

Index(['Country name', 'year', 'Life Ladder', 'Log GDP per capita',
       'Social support', 'Healthy life expectancy at birth',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'Positive affect', 'Negative affect',
       'Confidence in national government'],
      dtype='object')

In [6]:
whr_2021["Healthy life expectancy at birth"].mean()

65.36767285448276

Nein man kann nicht die durchschnittliche Lebenserwartung aller Menschen berechnen, da jedes Land eine verschieden große Population besitzt

WIr untersuchen nun für das Jahr 2021 folgende Frage:
>Gibt es einen Zusammenhang zwischen der Lebenserwartung ("Healthy life expectancy at birth") und der Lebenszufriedenheit ("Life Ladder") in einem Land? 

Um das herauszufinden, visualisieren wir die Daten. So wie man die Verteilung *einer* Variablen mit einem Histogramm oder einem Dichteplot visualisiert, haben Statistiker eine allgemeine Methode für die Visualisierung der Beziehung zwischen zwei numerischen Variablen entwickelt: das Streudiagramm. Das Streudiagramm (oder Scatterplot) wurde auch als die ["nützlichste Erfindung in der Geschichte der statistischen Grafik"](https://onlinelibrary.wiley.com/doi/abs/10.1002/jhbs.20078) bezeichnet. Es handelt sich um ein einfaches zweidimensionales Diagramm, bei dem die beiden Koordinaten jedes Punktes die Werte von zwei Variablen darstellen, die bei einer einzigen Beobachtung gemessen wurden.

Wir erzeugen wir einen Streuplot mit plotly mit folgendem Befehl:

In [7]:
px.scatter(data_frame=whr_2021, x="Healthy life expectancy at birth", y="Life Ladder")

Wenn Sie die Maus über das Diagramm bewegen, werden zu jedem Punkt die x- und y-Werte angezeigt - praktisch. Noch besser wäre es, wenn wir zusätzlich zu jedem Punkt sehen würden, welches Land sich dahinter verbirgt. Dies können wir erreichen, indem wir dem Befehl `px.scatter` über das Argument `hover_name=` den Spaltennamen des DataFrames übergeben, der zu jedem Punkt den Namen des Landes enthält,hier also: `Country name`:

In [8]:
px.scatter(data_frame=whr_2021, x="Healthy life expectancy at birth", y="Life Ladder", hover_name="Country name")

### ✏️ Aufgabe 4
In welchem Land ist die Lebenserwartung am geringsten, in welchem am höchsten? In welchem Land sind die Menschen am zufriedensten mit ihrem Leben, in welchem am unzufriedensten?

In [9]:
whr_2021.sort_values("Healthy life expectancy at birth").head(1)#am niedrigsten

Country name  year  Life Ladder  Log GDP per capita  Social support  \
1265   Mozambique  2021     5.178487            7.108412        0.664023   

      Healthy life expectancy at birth  Freedom to make life choices  \
1265                         51.299999                        0.8381   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  \
1265    0.048487                   0.627466         0.575795         0.383463   

      Confidence in national government  
1265                           0.731972

In [10]:
whr_2021.sort_values("Healthy life expectancy at birth",na_position="first").tail(1)#am höchsten

Country name  year  Life Ladder  Log GDP per capita  Social support  \
922        Japan  2021     6.091325           10.623948        0.895738   

     Healthy life expectancy at birth  Freedom to make life choices  \
922                         74.349998                       0.80135   

     Generosity  Perceptions of corruption  Positive affect  Negative affect  \
922   -0.209871                   0.669547         0.674374         0.189246   

     Confidence in national government  
922                           0.290786

In [11]:
whr_2021.sort_values("Life Ladder").head(1)#am unzufriedesten

Country name  year  Life Ladder  Log GDP per capita  Social support  \
1053      Lebanon  2021     2.178809            9.243084        0.506775   

      Healthy life expectancy at birth  Freedom to make life choices  \
1053                         66.150002                      0.422697   

      Generosity  Perceptions of corruption  Positive affect  Negative affect  \
1053   -0.142942                   0.905279         0.263487         0.569101   

      Confidence in national government  
1053                           0.106694

In [12]:
whr_2021.sort_values("Life Ladder",na_position="first").tail(1)#am zufriedesten

Country name  year  Life Ladder  Log GDP per capita  Social support  \
601      Finland  2021     7.794378           10.794553        0.970289   

     Healthy life expectancy at birth  Freedom to make life choices  \
601                         71.150002                       0.96319   

     Generosity  Perceptions of corruption  Positive affect  Negative affect  \
601   -0.035728                   0.191754         0.751612         0.175449   

     Confidence in national government  
601                           0.713731

Die Punkte im Streudiagramm liegen zwar verteilt in der Ebene, aber man kann doch erkennen, dass im Allgemeinen mit zunehmender Lebenserwartung die Lebenszufriedenheit zunimmt. Dies quantifizieren wir nun mit Hilfe des **Korrelationskoeffizienten**.

Es gibt unterschiedliche Möglichkeiten, diesen in Python zu berechnen. Wir nutzen die `.corr` Funktion von Pandas:

In [13]:
whr_2021["Healthy life expectancy at birth"].corr(whr_2021["Life Ladder"])

0.7262499344070562

Der Wert 0.726 besagt, dass es eine positive Korrelation zwischen Lebenserwartung und Lebenszufriedenheit gibt, dass also mit steigender Lebenserwartung auch die Lebenszufriedenheit steigt - und umgekehrt. Dieser Zusammenhang gilt nicht in jedem Einzelfall wie man leicht im Streudiagramms sieht. Eine gute Möglichkeit, sich die Stärke einer Korrelation zu veranschaulichen ist, die Regressionsgerade zu plotten. Würden alle Punkte genau auf der Geraden liegen, so hätte man eine perfekte Korrelation: +1, wenn die Gerade eine positive Steigung hat oder -1, wenn die Gerade eine negative Steigung hat (dies kommt bei realistischen Daten quasi nie vor).

plotly erlaubt es, die Regressionsgerade direkt im Plot zu visualisieren. Dafür muss das Modul `statsmodels` installiert sein (in der Anaconda Konsole `pip install statsmodels` eingeben). Die einfache Regressionsgerade wird über das Argument `trendline="ols"` aktiviert (ols steht für ordinary least squares).

In [14]:
px.scatter(data_frame=whr_2021, x="Healthy life expectancy at birth", y="Life Ladder", hover_name="Country name", trendline="ols")

Die positive Steigung der Geraden impliziert, dass eine *positive* Korrelation vorliegt. Die *Stärke* der Korrelation ist dadurch bestimmt, wie nahe an der Geraden die Punkte typischerweise liegen. Regressionsgeraden sind also dazu geeignet "Korrelationen sichtbar zu machen". 

Praktisch: Wenn Sie die Maus über die Gerade bewegen, wird die Geradengleichung, sowie das R² Gütmaß angezeigt. Zur Erinnerung: R² ist ein Wert zwischen 0 und 1, der aussagt, wie viel der Varianz in den Daten durch den Prediktor (hier: Lebenserwartung) erklärt wird. R² ist der quadrierte Korrelationskoeffizient.  

Wenn wir die Koeffizienten der Geraden nun automatisch auslesen möchten, können wir dies wie folgt tun. Wir speichern den *Plot* in einer Variable:

In [15]:
p = px.scatter(data_frame=whr_2021, x="Healthy life expectancy at birth", y="Life Ladder", hover_name="Country name", trendline="ols")

Wenn Sie diesen Code ausführen, werden Sie merken, dass der Plot nicht angezeigt wird (er wird erzeugt und in der Variable `p` gespeichert). Wir können die Koeffizienten $\alpha$ uund $\beta$ nun mittels folgenden Befehlen aus `p` extrahieren:

In [16]:
result = px.get_trendline_results(p)
result.iloc[0,0].params

array([-4.19493172,  0.15021072])

Übrigens: mit `p.show()` könnten Sie den Plot jederzeit anzeigen.

### ✏️ Aufgabe 5
Berechnen Sie die Koeffizienten $\alpha$ und $\beta$ der Regressionsgeraden $\hat{y}=\alpha+\beta x$ mittels der Formeln aus der Vorlesung und vergleichen Sie mit den Werten oben:
- $\beta = r_{xy}\frac{\tilde{s}_y}{\tilde{s}_x}$
- $\alpha = \bar{y} - \beta \bar{x}$

Welchen Wert für Lebenszufriedenheit sagt das Modell für eine Lebenserwartung von 80 Jahren voraus?

Annahme Stichprobenstandardabweichung ähnlich zur reallen standardabweichung

In [35]:
r = whr_2021["Healthy life expectancy at birth"].corr(whr_2021["Life Ladder"])
sy = whr_2021["Life Ladder"].std()
sx = whr_2021["Healthy life expectancy at birth"].std()
beta = r*sy/sx
alpha = whr_2021["Life Ladder"].mean()-beta*whr_2021["Healthy life expectancy at birth"].mean()

In [36]:
alpha

-4.106315849282506

In [37]:
beta

0.14902795280321104

Antwort sind sehr ähnliche werte mit ein bisschen abweichung

In [38]:
#für 80 Lebensjahre
alpha+beta*80

7.815920374974377

---
Der Korrelation zwischen Lebenserwartung und Zufriedenheit war ein Beispiel für eine positive Korrelation. Natürlich können Sie für jedes beliebige Paar von Variablen die Korrelation berechnen.

### ✏️ Aufgabe 6
Zwischen welchen Variablen im Datensatz `whr_2021` vermuten Sie eine *negative* Korrelation? Warum? Verifizieren Sie ihre Vermutung, indem Sie
1. den Korrelationskoeffizienten berechnen und
2. die beiden betreffenden Merkmale sowie die Regressionsgerade in einem Streuplot anzeigen.

Vermutung negative Korrelaton Feature Life Ladder und Perception pf corruption

In [17]:
whr_2021.columns

Index(['Country name', 'year', 'Life Ladder', 'Log GDP per capita',
       'Social support', 'Healthy life expectancy at birth',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'Positive affect', 'Negative affect',
       'Confidence in national government'],
      dtype='object')

In [18]:
px.scatter(data_frame=whr_2021, x="Perceptions of corruption", y="Life Ladder", hover_name="Country name", trendline="ols")

In [19]:
whr_2021["Perceptions of corruption"].corr(whr_2021["Life Ladder"])

-0.45276340739374815